feat: add node management gui jobs

herve-er opened the pull request:
**depend on #1212** 
This commit introduces several new jobs for enhanced node management:
- Added `NodeFolderSizeJob`, `NodeInfoJob`, `NodeSubFoldersJob`, `SyncNodeListJob`, and `SyncNodeSetListJob` for operations like folder size calculation, node info retrieval, and sync list updates.
- Integrated these jobs into `GuiJobFactory`.

Refactored `ServerRequests` to support these jobs:
- Added methods like `getNodeInfo`, `getSubFolders`, and `getFolderSize`.

Removed deprecated `propagateSyncListChange` logic:
- Deleted the method from `GuiRequests` and its references in `AppServer`.

Updated `CMakeLists.txt` to include new job files.

Added unit tests for the new jobs:
- Tests include `testNodeSubFolderJob`, `testNodeFolderSizeJob`, `testSyncNodeListJob`, `testSyncNodeSetListJob`, and `testNodeInfoJob`.

Performed code cleanup and optimizations:
- Replaced `QList` with `std::vector` for better performance.
- Improved thread safety with `std::scoped_lock`.

herve-er committed:
feat: add node management gui jobs

This commit introduces several new jobs for enhanced node management:
- Added `NodeFolderSizeJob`, `NodeInfoJob`, `NodeSubFoldersJob`,
  `SyncNodeListJob`, and `SyncNodeSetListJob` for operations like
  folder size calculation, node info retrieval, and sync list updates.
- Integrated these jobs into `GuiJobFactory`.

Refactored `ServerRequests` to support these jobs:
- Added methods like `getNodeInfo`, `getSubFolders`, and `getFolderSize`.

Removed deprecated `propagateSyncListChange` logic:
- Deleted the method from `GuiRequests` and its references in `AppServer`.

Updated `CMakeLists.txt` to include new job files.

Added unit tests for the new jobs:
- Tests include `testNodeSubFolderJob`, `testNodeFolderSizeJob`,
  `testSyncNodeListJob`, `testSyncNodeSetListJob`, and `testNodeInfoJob`.

Performed code cleanup and optimizations:
- Replaced `QList` with `std::vector` for better performance.
- Improved thread safety with `std::scoped_lock`.

luc-guyot-infomaniak commented on src/server/comm/guijobs/nodefoldersizejob.cpp:
```diff
@@ -0,0 +1,66 @@
+        readParamValue(inParamsDriveId, _driveId);
+        readParamValue(inParamsNodeId, _nodeId);
+    } catch (const std::exception &e) {
+        LOG_WARN(_logger, "Exception in AbstractGuiJob::readParamValue: error=" << e.what());
```
```suggestion
        LOG_WARN(_logger, "Exception in NodeFolderSizeJob::readParamValue: error=" << e.what());
```

herve-er replied:
I would tend to keep it as it is, as *readParamValue* is a function of *AbstractGuiJob*, not *NodeFolderSizeJob*.

Or, if we really want to avoid mentioning *AbstractGuiJob*, it should be:
`Exception in DriveDeleteJob::deserializeInputParms: error=`

luc-guyot-infomaniak replied:
The original wording is correct. My point is that mentioning the derived class name gives more context to the log message. (The method `readParamValue` is a method of the base class but also of every derived class).

herve-er replied:
Done here: https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/a439aa9e0cb1088f221d87eec51adc358e203cd9

luc-guyot-infomaniak commented on src/server/comm/guijobs/nodeinfojob.cpp:
```diff
@@ -0,0 +1,74 @@
+        readParamValue(inParamsNodeId, _nodeId);
+        readParamValue(inParamsWithPath, _withPath);
+    } catch (const std::exception &e) {
+        LOG_WARN(_logger, "Exception in AbstractGuiJob::readParamValue: error=" << e.what());
```
```suggestion
        LOG_WARN(_logger, "Exception in NodeInfoJob::readParamValue: error=" << e.what());
```

herve-er replied:
https://github.com/Infomaniak/desktop-kDrive/pull/1214#discussion_r2554925996

luc-guyot-infomaniak commented on src/server/comm/guijobs/nodesubfoldersjob.cpp:
```diff
@@ -0,0 +1,74 @@
+        readParamValue(inParamsNodeId, _nodeId);
+        readParamValue(inParamsWithPath, _withPath);
+    } catch (const std::exception &e) {
+        LOG_WARN(_logger, "Exception in AbstractGuiJob::readParamValue: error=" << e.what());
```
```suggestion
        LOG_WARN(_logger, "Exception in NodeSubFoldersJob::readParamValue: error=" << e.what());
```

herve-er replied:
https://github.com/Infomaniak/desktop-kDrive/pull/1214#discussion_r2554925996

luc-guyot-infomaniak commented on src/server/comm/guijobs/nodesubfoldersjob.cpp:
```diff
@@ -0,0 +1,74 @@
+
+ExitInfo NodeSubFoldersJob::serializeOutputParms() {
+    // Output parameters serialization
+    std::function<Poco::Dynamic::Var(const NodeInfo &)> nodeInfo2DynamicVar = [](const NodeInfo &value) {
```
This function and few others could be defined in a file named `serializerHelpers.cpp`.

herve-er replied:
I missed the helper already implemented by @ChristopheLarchier, simplified here: [[refactor: simplify object-to-dynamic conversion logic](https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/1447d31c32536ca60f93d98eb45a8daec49480df)](https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/1447d31c32536ca60f93d98eb45a8daec49480df)

luc-guyot-infomaniak commented on src/server/comm/guijobs/syncnodelistjob.cpp:
```diff
@@ -0,0 +1,78 @@
+        readParamValue(inParamsSyncDbId, _syncDbId);
+        readParamValue(inParamsSyncNodeType, _syncNodeType);
+    } catch (const std::exception &e) {
+        LOG_WARN(_logger, "Exception in AbstractGuiJob::readParamValue: error=" << e.what());
```
```suggestion
        LOG_WARN(_logger, "Exception in SyncNodeListJob::readParamValue: error=" << e.what());
```

herve-er replied:
https://github.com/Infomaniak/desktop-kDrive/pull/1214#discussion_r2554925996

luc-guyot-infomaniak commented on src/server/comm/guijobs/syncnodelistjob.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+ExitInfo SyncNodeListJob::serializeOutputParms() {
+    // Output parameters serialization
```
```suggestion
```
Because the name of the method is explicit enough.

herve-er replied:
Done here: https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/e50bc296996f72a4590c267b5b92d52af05b21c3

luc-guyot-infomaniak commented on src/server/requests/serverrequests.cpp:
```diff
@@ -780,12 +783,26 @@ ExitCode ServerRequests::addSync(int driveDbId, const QString &localFolderPath,
+    if (!exitInfo) {
+        return exitInfo;
+    }
+    list.clear();
```
This should be placed at the top.

herve-er replied:
Done here: https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/80f0503c0eaaf8433d4afeea01d3e37c8e4d63be

luc-guyot-infomaniak commented on test/server/comm/testguicommchannel.h:
```diff
@@ -95,6 +95,12 @@ class TestGuiCommChannel : public CppUnit::TestFixture, public TestBase {
         void testSyncStartAfterLoginJob();
         void testSyncDeleteJob();
         void testSyncGetPublicLinkUrlJob();
+        void testNodeSubFolderJob();
```
You forgot to submit the unit tests implementation.

herve-er replied:
I extracted those changes from a bigger PR, which brought the unit tests at risk of causing a lot of conflicts, as they also change the names of those jobs. I will include them in a future PR (just after [https://github.com/Infomaniak/desktop-kDrive/pull/1214](https://github.com/Infomaniak/desktop-kDrive/pull/1214)).

luc-guyot-infomaniak commented on src/libcommon/comm.h:
```diff
@@ -69,7 +69,6 @@ enum class RequestNum {
     SYNC_ASKFORSTATUS,
     SYNC_SETSUPPORTSVIRTUALFILES,
     SYNC_SETROOTPINSTATE,
-    SYNC_PROPAGATE_SYNCLIST_CHANGE,
```
There are other instances in the code, e.g., in`desktop-kDrive/src/front/macOS/kDriveCore/ServerBridge/XPC/comm.h`.

luc-guyot-infomaniak commented on src/server/appserver.cpp:
```diff
@@ -1519,7 +1519,11 @@ void AppServer::onRequestReceived(int id, RequestNum num, const QByteArray &para
                 addError(Error(ERR_ID, exitCode, ExitCause::Unknown));
             }
-
+            exitCode = syncPalMapIt->second->syncListUpdated(true);
```
As `GuiRequests::setSyncIdSet` is called three times in `DrivePreferencesWidget::updateSelectiveSyncList`, we are going to queue three blacklist propagator jobs after a single change in the folder tree (GUI).

herve-er committed:
chore(comment): Remove redundant comments.

herve-er committed:
fix: Ensure getUserInfoList return an empty list in case of error.

herve-er committed:
refactor: simplify object-to-dynamic conversion logic

Refactored `serializeOutputParms()` methods in `NodeInfoJob`,
`NodeSubFoldersJob`, and `UserAvailableDrivesJob` to replace
redundant lambda functions with the generic `info2DynamicVar<T>`
utility. This change reduces code duplication and improves
maintainability.

herve-er committed:
fix: improve logging and clean up unused attributes

Refactor logging messages for better clarity and accuracy:
- Updated error message in `appserver.cpp` to reflect the correct method causing the issue.
- Reformatted logging in `serverrequests.cpp` to improve readability and fix formatting inconsistencies.

Code cleanup:
- Removed unused `_withPath` attribute from `NodeFolderSizeJob` in `nodefoldersizejob.h`.
- Changed `_requestNum` value in `SyncNodeSetListJob` constructor in `syncnodesetlistjob.cpp` to align with the correct request type.

Minor formatting fixes:
- Corrected spacing in `getFolderSizeWithCallback` method in `serverrequests.cpp`.

herve-er committed:
fix(logging): improve class-specific log messages

Replaced generic `AbstractGuiJob` references in `LOG_WARN`
messages with the specific class names in multiple files.
This change enhances the precision and traceability of
log messages, making it easier to identify the source of
exceptions during execution.

No functional or logical changes were made; the updates
are limited to improving log clarity.

luc-guyot-infomaniak commented on src/server/comm/guijobs/nodesubfoldersjob.cpp:
```diff
@@ -0,0 +1,67 @@
+}
+
+ExitInfo NodeSubFoldersJob::process() {
+    return ServerRequests::getSubFolders(_userDbId, _driveId, _nodeId, _nodeSubFolderInfoList, _withPath);
```
You have decided not to reproduce the log message and the call to `AppServer::addError` from  `appserver.cpp`.
What is the rational behind this removal?

herve-er replied:
Done here: https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/5f0573f9f7728682a30084e242b4ae530ef70eaf

~~However, I'm doubtful about the usefulness of the double error reporting (exitInfo and addError) for such a sync-unrelated job.~~ -> It might indeed be usefull to detect token expiration early

herve-er committed:
fix(error-handling): improve robustness in job processing

Enhanced error handling across multiple job classes:
- Updated `NodeInfoJob::process()` to log warnings and handle errors from `ServerRequests::getNodeInfo`.
- Improved `NodeSubFoldersJob::process()` with similar error handling for `ServerRequests::getSubFolders`.
- Added error logging and handling in `SyncNodeListJob::process()` for `syncIdSet` failures.
- Enhanced `SyncNodeSetListJob::process()` with error checks for `setSyncIdSet` and `syncListUpdated`.

luc-guyot-infomaniak commented on src/server/comm/guijobs/syncnodesetlistjob.h:
```diff
@@ -0,0 +1,42 @@
+        std::vector<NodeId> _nodeIdList;
+
+        ExitInfo deserializeInputParms() override;
+        ExitInfo serializeOutputParms() { return ExitCode::Ok; }
```
```suggestion
        ExitInfo serializeOutputParms() override { return ExitCode::Ok; }
```

herve-er replied:
Done here: https://github.com/Infomaniak/desktop-kDrive/pull/1214/commits/049af36290365db6918ae730b49789a25a3e7086

herve-er committed:
Update src/server/comm/guijobs/syncnodesetlistjob.h

Co-authored-by: Luc Guyot <162997198+luc-guyot-infomaniak@users.noreply.github.com>
